### PR TITLE
Fix bug when num_pos are different in a batch

### PR DESCRIPTION
**Pull request by Sundrops**

### CODE REVIEW

File: det3d/models/losses/centernet_loss.py [hunk 1]
```diff
@@ -19,8 +19,8 @@ def forward(self, output, mask, ind, target):
     mask = mask.float().unsqueeze(2) 
 
     loss = F.l1_loss(pred*mask, target*mask, reduction='none')
-    loss = loss / (mask.sum() + 1e-4)
-    loss = loss.transpose(2 ,0).sum(dim=2).sum(dim=1)
+    loss = loss / (mask.sum(dim=(1, 2)).unsqueeze(1).unsqueeze(2) + 1e-4)
+    loss = loss.transpose(2, 0).sum(dim=2).sum(dim=1)
```
Comment (tianweiy) [inline det3d/models/losses/centernet_loss.py:23]: change to mean(dim=2) otherwise, the loss scale is dependent on the batch size

File: det3d/models/losses/centernet_loss.py [hunk 2]
```diff
@@ -45,10 +45,12 @@ def forward(self, out, target, ind, mask, cat):
```
Comment (tianweiy) [inline det3d/models/losses/centernet_loss.py:45]: change to neg_loss = neg_loss.mean(dim=(1, 2, 3))

File: det3d/models/losses/centernet_loss.py [hunk 3]
```diff
@@ -45,10 +45,12 @@ def forward(self, out, target, ind, mask, cat):
...
-    pos_loss = pos_loss.sum()
-    if num_pos == 0:
-      return - neg_loss
-    return - (pos_loss + neg_loss) / num_pos
+    pos_loss = pos_loss.sum(dim=(1, 2))
+    mask = (num_pos > 0).float()   # ind of num_pos > 0
+    lossA = - neg_loss
+    lossB = - (pos_loss + neg_loss) / (num_pos + 1e-4)
+    loss = mask * lossB + (1 - mask) * lossA
+    return loss.sum()
```
Comment (tianweiy) [inline det3d/models/losses/centernet_loss.py:56]: mean() same as comment above

**Review by tianweiy — COMMENTED** [review]
Thanks for the contribution. I added a few comments

File: det3d/models/losses/centernet_loss.py [hunk 4]
```diff
@@ -19,8 +19,9 @@ def forward(self, output, mask, ind, target):
     mask = mask.float().unsqueeze(2) 
 
     loss = F.l1_loss(pred*mask, target*mask, reduction='none')
-    loss = loss / (mask.sum() + 1e-4)
-    loss = loss.transpose(2 ,0).sum(dim=2).sum(dim=1)
+    loss = loss / (mask.sum(dim=(1, 2)).unsqueeze(1).unsqueeze(2) + 1e-4)
+    # loss: B * max_objects * dim(x,y,z,l,w,h,sin,cos)
+    loss = loss.transpose(2, 0).mean(dim=2).mean(dim=1)
```
Comment (tianweiy) [inline det3d/models/losses/centernet_loss.py:24]: no need for second mean(dim=1) ? we can do sum there as we already divide by num_pos ?

**Review by tianweiy — COMMENTED** [review]
just one comment. The other looks good

File: det3d/models/losses/centernet_loss.py [hunk 5]
```diff
@@ -16,12 +16,13 @@ def __init__(self):
...
   def forward(self, output, mask, ind, target):
     pred = _transpose_and_gather_feat(output, ind)
+    num_pos = mask.sum(dim=1)
     mask = mask.float().unsqueeze(2) 
 
     loss = F.l1_loss(pred*mask, target*mask, reduction='none')
     loss = loss / (mask.sum(dim=(1, 2)).unsqueeze(1).unsqueeze(2) + 1e-4)
     # loss: B * max_objects * dim(x,y,z,l,w,h,sin,cos)
-    loss = loss.transpose(2, 0).mean(dim=2).mean(dim=1)
+    loss = loss.transpose(2, 0).sum(dim=2).mean(dim=1)/num_pos
```
Comment (tianweiy) [inline det3d/models/losses/centernet_loss.py:25]: just sum, no need to divide?  we already divide by num_pos in line 23.

**Comment by tianweiy** [comment]
thanks again for the contribution. I trained one model but it is slightly worse than the original version. I will look into this further.

**Comment by Sundrops** [comment]
Maybe it's need to increase lr
